### PR TITLE
Fix PathTemplate matcher to match path with protocol and hostname

### PR DESCRIPTION
**Pull request by ajaaym**
Fixes [#3604](https://github.com/googleapis/google-cloud-java/issues/3604) Fix matcher to match path with protocol and hostname

### CODE REVIEW

**Comment by andreamlin** [comment]
Thanks for filing this PR!

Does this PR address anything else other than https://github.com/googleapis/google-cloud-java/issues/3604? 

Regarding https://github.com/googleapis/google-cloud-java/issues/3604, can you write a test that parses ` https://www.googleapis.com/compute/v1/projects/project-123/zones/europe-west3-c`? I'm not sure if this code would be able to reconcile the `compute/v1` substring between the PathTemplate string and the hostname.

(With regard to that issue, our plan is to strip off the `https://www.googleapis.com/compute/v1/projects` string, defined in the baseUrl of the compute.v1.json API definition file from which the Java Compute client is generated, from the returned path, so that we are left with a parseable `project-123/zones/europe-west3-c` substring that matches a path template `{project}/zones/{zone}/`, also defined in the compute.v1.json doc.)

**Comment by ajaaym** [comment]
@andreamlin Thanks for review. Added test case for  [googleapis/google-cloud-java#3604](https://github.com/googleapis/google-cloud-java/issues/3604). This PR was suppose to be precursor to fix [googleapis/google-cloud-java#3604](https://github.com/googleapis/google-cloud-java/issues/3604) as well as in general to support the url with protocol and hostname. Once this PR gets merged, need to fix Path template string for ProjectZoneName to fix [googleapis/google-cloud-java#3604](https://github.com/googleapis/google-cloud-java/issues/3604). This is non breaking change would also work with your other solution.

File: src/main/java/com/google/api/pathtemplate/PathTemplate.java [hunk 1]
```diff
@@ -132,6 +132,9 @@
   // A regexp to match a custom verb at the end of a path.
   private static final Pattern CUSTOM_VERB_PATTERN = Pattern.compile(":([^/*}{=]+)$");
 
+  // A regex to match a hostname with or without protocol.
+  private static final Pattern HOSTNAME_PATTERN = Pattern.compile("^(\\w+:)?//");
```
Comment (andreamlin) [inline src/main/java/com/google/api/pathtemplate/PathTemplate.java]: Now that `HOSTNAME_PATTERN` also matches with a protocol, can you update the comment for `HOSTNAME_VAR` to explain that that might also contain a protocol string?

**Comment by ajaaym** [comment]
@andreamlin can you please review?

**Pull request merged**